File: leetcode/56_merge_intervals.cpp
struct compare{
    bool operator() (pair<int, int> a, pair<int, int> b){
         return !(a.second > b.second);
    }
};

struct compare2 {
    bool operator() (vector<int> a, vector<int> b) {
         if (a[0] != b[0]) {
             return !(a[0] > b[0]);
         } if (a[1] != b[1]) {
             return !(a[1] < b[1]);
         }
         return false;
    }
};

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        int n = intervals.size();
        sort(intervals.begin(), intervals.end(), compare2());
        priority_queue<pair<int, int>, vector<pair<int, int>>, compare> q;
        vector<vector<int>> result;
        q.push({intervals[0][0], intervals[0][1]});
        for(int i=1; i<n; i++) {
            if (!q.empty() && intervals[i][0] <= q.top().second) {
                if (intervals[i][1] > q.top().second) {
                    pair<int, int> curr = q.top();
                    q.pop();
                    q.push({curr.first, intervals[i][1]});
                }
            } else {
                q.push({intervals[i][0], intervals[i][1]});
            }
        }
        while (!q.empty()) {
            pair<int, int> curr = q.top();
            q.pop();
            vector<int> t = vector<int>{curr.first, curr.second};
            result.push_back(t);
        }
        return result;
    }
};

File: leetcode/56_merge_intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return intervals
        intervals.sort(key=lambda item: item[0])
        START = 0
        END = 1
        curr_max_start = intervals[0][START]
        curr_max_end = intervals[0][END]
        result = []
        for i in range(1, len(intervals)):
            if intervals[i][START] > curr_max_end:
                result.append([curr_max_start, curr_max_end])
                curr_max_start = intervals[i][START]
                curr_max_end = intervals[i][END]
            else:
                curr_max_end = max(curr_max_end, intervals[i][END])
        result.append([curr_max_start, curr_max_end])
        return result
